### ISSUE TITLE
Some updates do not seem to be syncing to binsync

### ISSUE THREAD

**Issue by Lukas-Dresel** [angr/binsync#124]
Open `phonebook` in IDA, go to `choose_relationship` @ 0x1a8b, from there click on `business_call` to highlight it (do not enter the function), hit `Y` to change the type. This does not show up in `binsync` for me.

**Comment by Lukas-Dresel** [comment]
This worked 1/3 times, the other times it did not show up. I made multiple changes (e.g. changing to a different type, then changing back), but they still did not show up. Is there some time delay between action + revert before it gets picked up? e.g. i changed the type, and change it back within a second. Could this be why it does not show up?

**Comment by mahaloz** [comment]
@Lukas-Dresel yeah it can be around 5 seconds before the change gets picked up for committing. After that, it can be another 10 seconds before you see in your GUI. Total trip: 15 seconds max.

**Comment by Lukas-Dresel** [comment]
huh okay
let me try with waiting in between then

**Comment by Lukas-Dresel** [comment]
Okay, yeah, still does not appear to sync it for me, even after waiting 20+ seconds, so definitely a bug

**Comment by Lukas-Dresel** [comment]
Hmm, so it's a bit weirder than that? If i set the argument type to `void*` it does not appear to sync? but if i set it to `person_t*` then it does? that makes no sense

**Comment by Lukas-Dresel** [comment]
@mahaloz Oh okay, so here's what it seems to be

It does not update when the type is changed. It only updates once you actually enter the changed function in the decompiler view!

**Comment by Lukas-Dresel** [comment]
I did the above to change the type, waited for a minute or so just watching the repo to see if new commits show up. They did not.

But the second I then entered the function manually, a change showed up.

**Comment by mahaloz** [comment]
@Lukas-Dresel wait wait wait, you changed a type in a function in IDA without actually being in the function?

**Comment by Lukas-Dresel** [comment]
i changed the function signature at the call site

**Comment by Lukas-Dresel** [comment]
just hit `y` on the function name

**Comment by Lukas-Dresel** [comment]
I changed the argument type.

**Comment by Lukas-Dresel** [comment]
@mahaloz 
![gif](https://user-images.githubusercontent.com/13377119/163761641-92ba7c58-4e88-4f60-81d4-fed92c3f76c8.gif)

**Comment by mahaloz** [comment]
## Problem
If you take a look at @Lukas-Dresel GIF above, you will notice that he changes the prototype of a function without actually being in that function. With the current way we hook decompilation, its not possible to change this. The only way to fix this behavior is to find some new hook in ida that allows us to catch prototype changes, which we have still not been able to find.

**Comment by arizvisa** [comment]
```
        ui.hook.idp.add('ev_init', typeinfo.database_init, 0)
        ui.hook.idb.add('changing_ti', typeinfo.changing, 0)
        ui.hook.idb.add('ti_changed', typeinfo.changed, 0)
```

**Comment by mahaloz** [comment]
@arizvisa Oh shnappers. Does the hook pass a reference to the actual object being changed? Or is just saying literally any object with a type is changing?

**Comment by arizvisa** [comment]
p/arameters for `changing_ti` (before changing the type, which requires you to extract the old type) are:
```
def changing_ti(ea :identifier, new_type :tinfo_t, new_name :str)`
```
parameters for `ti_changed` (after the "new_type" has been applied) are:
```
def ti_changed(ea :identifier, result_type :tinfo_t, fnames :maybe(str?)`
```
So you're going to need to correlate these two callbacks to determine the source type being changed in "changing_ti", to when the type __actually__ gets applied in "ti_changed". There's a few of the event hooks that use this changing/changed detmethodology. 

The "ea" parameter (your "reference") is always an integer "id" or "identifier". which is really just an address that references an address in the database..or is an identifier to the "structure". To distinguish these, I'll use "address" to refer to an address that's visible in the database and thus associated with a function or other visible address, and "identifier" to refer to an address that's actually one of the private addresses, that are used for structures, enumbers, etc. and are hence non-visible.

In most cases, it's okay to treat it like an address (by comparing it against the bounds of the database) to distinguish from an identifier. Iff you plan on using that identifier for xrefs or tracking changes for the types/structures themselves (again, *not* prototypes), keep in mind that all the private identifiers can be translated when rebasing the database. (I believe this to be out-of-scope for binsync, since your goals are just for monitoring `tinfo_t` which means you're just going to want to get the name of the typeinfo to be able snag it from `til_t` whenever you determine the reference is not an address).

The "proper" (and probably most robust) way to distinguish the integer reference (ea) between whether it's associated with an "address" or an "identifier", is by checking the reference using __all__ of the identifier APIs (`idaapi.get_struc`, `idaapi.get_enum_idx`, etc.). However, if you only care about determining whether it's an "address", I believe that checking against the database address boundaries should be fine. Although, there's also a way by comparing the integer against one of the fields in `idainfo`, namley `INF_PRIVRANGE_*`, if you want to distinguish an address from a private identifier.

If you end up deciding to handle the events for __both__ identifiers *and* addresses, make sure you do a rebase so you can test them before and after rebasing.
